[DDP] Support for multiple backwards

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59359 [DDP] Support for multiple backwards**

Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

ci-all version of this PR: https://github.com/pytorch/pytorch/pull/59931

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

rohan-varma committed:
[DDP] Support for multiple backwards



Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move prepare_for_backward into _DDPSink backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with retain_graph=True.
Also includes a fix for static graph training that is described in https://github.com/pytorch/pytorch/issues/58111.

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move prepare_for_backward into _DDPSink backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with retain_graph=True.
Tests are added for DDP regular training (non-static graph, non find unused parameters) non-static graph with unused parameters, and static graph without/without find unused parameters.
Also includes a fix for static graph training that is described in https://github.com/pytorch/pytorch/issues/58111.

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move prepare_for_backward into _DDPSink backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with retain_graph=True.
Tests are added for DDP regular training (non-static graph, non find unused parameters) non-static graph with unused parameters, and static graph without/without find unused parameters.
Also includes a fix for static graph training that is described in https://github.com/pytorch/pytorch/issues/58111.

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -840,37 +845,30 @@ def forward(self, *inputs, **kwargs):
+        state_dict = {
+            'static_graph': self.static_graph,
+            'find_unused': self.find_unused_parameters,
+            'num_iterations': self.num_iterations,
```
self.num_iterations is calculated for forward pass in python side, may not be synced with num_bwd_calls_ in C++

rohan-varma replied:
Agreed it is not synced, but we really only use this to decide whether we need to enqueue delay allreduce right? And we only enqueue delay allreduce exactly once (basically only the first call to backward), so just checking num_iterations == 1 should be sufficient as delay allreduce will run in the next bwd.

Although agreed it might be a bit cleaner to directly check the num_backward_calls, perhaps

rohan-varma replied:
discussed offline, we are using  ctx.reduer.get_num_backward_calls_() as suggested

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -130,18 +130,23 @@ def backward(ctx, *grad_outputs):
         if static_graph_training and ctx.state_dict['num_iterations'] == 1:
             Variable._execution_engine.queue_callback(ctx.reducer._delay_all_reduce)
```
this may need to be moved after prepare_for_backward() call, and use ctx.reduer.get_num_backward_calls_() to check whether it is first backward iteration

rohan-varma replied:
Would it be simpler if we just have a boolean flag in python associated with static graph training? We always only need to enqueue it once, and only the first time, so something like `enqueued_delay_allreduce` flag should be enough and simpler here?

rohan-varma replied:
Also if there are N forward calls before a backward, checking `num_backward_calls` in from C++ may result in us enqueueing this multiple times, which is not wanted

rohan-varma replied:
discussed offline, we are using  ctx.reduer.get_num_backward_calls_() as suggested

zhaojuanmao reviewed (commented):
overall looks good! left one comment

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move prepare_for_backward into _DDPSink backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with retain_graph=True.
Tests are added for DDP regular training (non-static graph, non find unused parameters) non-static graph with unused parameters, and static graph without/without find unused parameters.
Also includes a fix for static graph training that is described in https://github.com/pytorch/pytorch/issues/58111.

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move prepare_for_backward into _DDPSink backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with retain_graph=True.
Tests are added for DDP regular training (non-static graph, non find unused parameters) non-static graph with unused parameters, and static graph without/without find unused parameters.
Also includes a fix for static graph training that is described in https://github.com/pytorch/pytorch/issues/58111.

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move prepare_for_backward into _DDPSink backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with retain_graph=True.
Tests are added for DDP regular training (non-static graph, non find unused parameters) non-static graph with unused parameters, and static graph without/without find unused parameters.
Also includes a fix for static graph training that is described in https://github.com/pytorch/pytorch/issues/58111.

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

zhaojuanmao reviewed (approved):
looks great, thanks for adding this!

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move prepare_for_backward into _DDPSink backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with retain_graph=True.
Tests are added for DDP regular training (non-static graph, non find unused parameters) non-static graph with unused parameters, and static graph without/without find unused parameters.
Also includes a fix for static graph training that is described in https://github.com/pytorch/pytorch/issues/58111.

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

zhaojuanmao commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1486,6 +1495,13 @@ void Reducer::finalize_backward() {
+  reset_bucket_counting();
+  // If we populated rebuilt params list in this backward call, avoid
+  // repopulating in subsequent backward calls.
+  all_rebuilt_params_pushed_ = !rebuilt_params_.empty();
```
rebuilt_params_() will be cleared after rebuild_bucket(), in this case, all_rebuilt_params_pushed_ will be true in every iteration after rebuilding bucket, and turns out to push rebuilt params many times, right?

rohan-varma replied:
Good point, although pushing rebuilt params is also guarded by `should_rebuild_bucket()` call, which will return false after we have rebuilt buckets, so this wouldn't cause an issue from what I can tell.

Although, it's probably better to change this to:
```
all_rebuilt_params_pushed = all_rebuilt_params_pushed || !rebuilt_params_.empty()

``

zhaojuanmao commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1486,6 +1495,13 @@ void Reducer::finalize_backward() {
 
+  // Reset bucket counting to ensure we can appropriately launch allreduce for
+  // each bucket in the next backwards.
+  reset_bucket_counting();
```
why this needs to be moved out of prepare_backward() and called at multiple places instead?

rohan-varma replied:
Discussed offline, added a helper function to call this at finalize_backward() and rebuild_buckets() where it is required.

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

ci-all version of this PR: https://github.com/pytorch/pytorch/pull/59931

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

ci-all version of this PR: https://github.com/pytorch/pytorch/pull/59931

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

ci-all version of this PR: https://github.com/pytorch/pytorch/pull/59931

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

ci-all version of this PR: https://github.com/pytorch/pytorch/pull/59931

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -142,23 +142,35 @@ def forward(ctx, reducer, state_dict, *inputs):
+        # static graph training as prepare_for_backward sets the
+        # num_backwards_call counter in the reducer.
+        static_graph_first_bwd = (
+            static_graph_training and ctx.reducer._static_graph_first_bwd()
```
nit: I'm wondering why static_graph_training is needed here? _static_graph_first_bwd() has already checked this?

rohan-varma replied:
If we are not in static graph mode at all, it would save one Python -> C++ -> Python round trip. I feel like that might be worth it as a small optimization in non-static graph mode.

zhaojuanmao reviewed (commented):
looks good! just one minor question

rohan-varma committed:
Update on "[DDP] Support for multiple backwards"


Move `prepare_for_backward` into `_DDPSink` backward instead of calling it in DDP forward pass so that we can run multiple backwards in DDP with `retain_graph=True`. 

ci-all version of this PR: https://github.com/pytorch/pytorch/pull/59931

Differential Revision: [D28855226](https://our.internmc.facebook.com/intern/diff/D28855226/)

[ghstack-poisoned]

rohan-varma commented:
Ci errors are unrelated, landing
